concordium-consensus/src/Concordium/External/DryRun.hs

```diff
@@ -0,0 +1,881 @@
+        return OK
+  where
+    ca = ContractAddress (ContractIndex contractIndex) (ContractSubindex contractSubindex)
+
+-- | Invoke an entrypoint on a smart contract instance in the current dry-run state.
+-- No changes to the state are retained at the completion of this operation.
+dryRunInvokeInstance ::
+    StablePtr DryRunHandle ->
+    -- | Contract index.
+    Word64 ->
+    -- | Contract subindex.
+    Word64 ->
+    -- | Invoker address tag.
+    Word8 ->
+    -- | Invoker account address pointer.
+    Ptr Word8 ->
+    -- | Invoker contract index.
+    Word64 ->
+    -- | Invoker contract subindex.
+    Word64 ->
+    -- | Amount.
+    Word64 ->
+    -- | ReceiveName pointer.
+    Ptr Word8 ->
+    -- | ReceiveName length.
+    Word32 ->
+    -- | Parameter pointer.
+    Ptr Word8 ->
+    -- | Parameter length.
+    Word32 ->
+    -- | Energy
+    Word64 ->
+    -- | Output vector.
+    Ptr ForeignVec ->
+    -- | The return code (see 'DryRunReturnCode').
+    IO Int64
+dryRunInvokeInstance
+    dryRunPtr
+    contractIndex
+    contractSubindex
+    invokerAddressType
+    invokerAccountAddressPtr
+    invokerContractIndex
+    invokerContractSubindex
+    amount
+    receiveNamePtr
+    receiveNameLen
+    parameterPtr
+    parameterLen
+    energy
+    outVec = dryRunStateHelper dryRunPtr outVec costInvokeInstanceBase $ \StateHelperInfo{..} -> do
+        maybeInvoker <- case invokerAddressType of
+            0 -> return Nothing
+            1 -> Just . AddressAccount <$> decodeAccountAddress invokerAccountAddressPtr
+            _ ->
+                return $
+                    Just $
+                        AddressContract $
+                            ContractAddress
+                                (ContractIndex invokerContractIndex)
+                                (ContractSubindex invokerContractSubindex)
+        method <- decodeReceiveName receiveNamePtr receiveNameLen
+        parameter <- decodeParameter parameterPtr parameterLen
+
+        let (energyLimit, quotaLimiting)
+                | Energy energy > shiQuotaRem = (shiQuotaRem, True)
+                | otherwise = (Energy energy, False)
+        let context =
+                InvokeContract.ContractContext
+                    { ccInvoker = maybeInvoker,
+                      ccContract =
+                        ContractAddress
+                            (ContractIndex contractIndex)
+                            (ContractSubindex contractSubindex),
+                      ccAmount = Amount amount,
+                      ccMethod = method,
+                      ccParameter = parameter,
+                      ccEnergy = energyLimit
+                    }
+        res <- runWithEBlockStateContext shiMVR shiBSC $ \drsRef -> do
+            -- We "freeze" the block state using a dummy hash so that we do not recompute the
+            -- state hash, since it is not required by invokeContract.
+            DryRunState{..} <- liftIO $ readIORef drsRef
```

Comment:
Could we change the InvokeContractMonad to not need the frozen  state instead?

Reply:
This should be part of a larger refactor to change how operations work on updatable/non-updatable block states.